### PR TITLE
163 implement the dal

### LINKED ISSUE #163: Implement the DAL
Last iteration we created a Data Layer Abstraction (DAL) design. Now we need to implement it to abstract our data storage interaction
https://app.diagrams.net/#G1pue7o1hbV5-o876cVSWB4kSA5G7GOVA3#%7B%22pageId%22%3A%22CqYIRr6Jr4gl-h0RM-nM%22%7D

### PR DESCRIPTION
**Pull request by LongQuanNguyen**
Closes #163 

# What was the issue

We did not have a layer to abstract away our data between how it was loaded and eventually used by the Graph components. Here is the  [DAL](https://github.com/UniversityOfSaskatchewanCMPT371/term-project-2024-team-2/wiki/Architecture-Design-Records-(ADRs)#5-data-access-layer-dal-abstraction-design) visually.

# What was done and why

The Data Abstraction Layer (DAL) accomplishes this by providing a simple interface between the Repository (IndexedDB) and the interactive Graph.

All the relevant dev files, listed as in directories:

**1. src/data:** 
- DataAbstractor.tsx (interface)
- DataLayer.tsx (main)

**2. src/repository:**
- Column.tsx
- DbRepository.tsx
- Repository.tsx (interface)

# How it was tested

All the current implemented functionalities have tests in place to verify and show how the IF works.
The relevant test files, listed as in directories:

**1.  test/data:**
- DataLayer.test.tsx
- PrivilegedDataLayer.tsx (Test Hooks)

**2. test/ repository:**
- DBRepository.test.tsx

### CODE REVIEW

**Comment by LongQuanNguyen** [comment]
`clear && npm run test -- DataLayer.test.tsx
`
to quickly run a test file in isolation.

File: my-webxr-app/tests/repository/DbRepository.test.tsx
Comment (Jesster2829) [inline my-webxr-app/tests/repository/DbRepository.test.tsx:1]: good test cases

**Review by Jesster2829 — APPROVED** [review]

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:119]: Yeah i also think the same, i try to adhere to best practices

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:15]: I think we keep this for the scope of one file

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (Aesha7) [inline my-webxr-app/src/repository/DbRepository.tsx:269]: Do both functions, getNumericColumnNames() and getAllColumnNames(), resolve to Promise? Are both called simultaneously or is one overwritten?

File: my-webxr-app/src/data/DataLayer.tsx
Comment (Aesha7) [inline my-webxr-app/src/data/DataLayer.tsx:125]: Good work from Tony's math library!

**Review by Aesha7 — APPROVED** [review]

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/repository/DbRepository.tsx:269]: Not sure if i got the question. 

Get allColumnNames return all the column names in the raw table -- all the header in the original headers in csv.

Get numericColumnNames return all the column names in the stat table, which are all numbers, i plan to use this for user to select columns to perform pca on.

Comment (LongQuanNguyen) [reply my-webxr-app/src/repository/DbRepository.tsx:269]: They both resolve to promise which are both array of strings

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:125]: Yippee! 🙌

File: my-webxr-app/tests/repository/DbRepository.test.tsx
Comment (trangnguyen3010) [inline my-webxr-app/tests/repository/DbRepository.test.tsx:1]: will be nice to add tests for adding columns to each table (since the column type is different for each case) and those stat calculation methods

File: my-webxr-app/src/data/DataLayer.tsx
Comment (trangnguyen3010) [inline my-webxr-app/src/data/DataLayer.tsx:1]: Loving all the functions' specifications! 💯

File: my-webxr-app/src/data/DataAbstractor.tsx
Comment (trangnguyen3010) [inline my-webxr-app/src/data/DataAbstractor.tsx:7]:
```suggestion
  storeCSV: (batchItems: BatchedDataStream) => Promise<boolean>;
```

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:15]: Bad communication up there sorry. Yes, i will work on this after everything get merged in

Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:15]: Hold up, if we use that option, then the batches will no longer be 2D arrays and cannot be transposed. That option return objects of key value pairs.

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:205]: Something not adding up here, all the test is passed... i will check asap

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/repository/DbRepository.tsx:252]: Oh good catch, that could be issue. Will look into this

File: my-webxr-app/src/data/DataAbstractor.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataAbstractor.tsx:7]: yes, I will fix this. Thanks so much!!!

Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataAbstractor.tsx:7]: Ahh, we run into dependency cycle with this.  DataLayer depends on  DataAbstracctor and DataAbstractor depends on BatchedDataStream from Datalayer. Will eslint disable it now

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:119]: right, forgot about the new type enforce, will fix this now

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:205]: oh just syntax thing, it still works. Those lines equivalent to:
```
const mean = statsColumn.values.mean;
const stdDev = statsColumn.values.stdDev;
```

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:187]: fixed

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:167]: fixed

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:211]: I see, my be define another type for column?  I am not sure how to go for this, can u take a look at Column.tsx and let me know where should I go form there?

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/repository/DbRepository.tsx:252]: I use Dexie's methods to refactor this into a beautiful one-liner.  Should have read the document before doing anything haha.
`
async getAllColumnNames(): Promise<string[]> {
    return await this.rawColumns.toCollection().keys() as string[];
  }
`

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx:187]: TODO added

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/repository/DbRepository.tsx:231]: I think so yeah. @Subzeero, what were your thoughts and intentions for that 231 line?

File: my-webxr-app/tests/repository/DbRepository.test.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/tests/repository/DbRepository.test.tsx:1]: D̶e̶l̶e̶g̶a̶t̶e̶d̶ t̶o̶ T̶r̶a̶n̶g̶

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (Aesha7) [reply my-webxr-app/src/repository/DbRepository.tsx]: Oh ok, thanks for clarifying!

File: my-webxr-app/src/data/DataLayer.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/data/DataLayer.tsx]: I made some changes to accommodate thís let me know what you think

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/repository/DbRepository.tsx]: Fixed

**Comment by LongQuanNguyen** [comment]
This PR is still a work in progress, I will merge it after @trangnguyen3010 reviews so that the rest of the dev team can complete this simultaneously

**Comment by LongQuanNguyen** [comment]
@MitchWag01 @qpb948 @Subzeero it would be nice if you guys could place some eyes on this.

File: my-webxr-app/src/data/DataLayer.tsx
Comment (trangnguyen3010) [inline my-webxr-app/src/data/DataLayer.tsx:330]: minor: I think BatchedDataStream makes more sense to be in CsvUtils as batches are generated in CsvUtils. That way we can avoid dependency cycle

**Review by trangnguyen3010 — APPROVED** [review]

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (trangnguyen3010) [inline my-webxr-app/src/repository/DbRepository.tsx:1]: might be beneficial to touch base with @qpb948 about changes in this PR and how they will impact his PR.

**Review by trangnguyen3010 — APPROVED** [review]

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (LongQuanNguyen) [reply my-webxr-app/src/repository/DbRepository.tsx:1]: yessssss!

**Comment by LongQuanNguyen** [comment]
I have implemented up to storeCSV function, with tests in place. The two main functions left are to select a PCA column and to select a raw data column for graphing. These two may overlap with @qpb948's card.

I will merge this in so that we can start working on hooking up stuff.

Please feel free to add more test cases as you see fit. Thanks.

File: my-webxr-app/src/data/DataAbstractor.tsx
Comment (Subzeero) [reply my-webxr-app/src/data/DataAbstractor.tsx:7]: In this case, we may be able to break out this type into a new file in the `types` folder. Something we can evaluate in the future.

File: my-webxr-app/src/data/DataLayer.tsx
Comment (Subzeero) [reply my-webxr-app/src/data/DataLayer.tsx:394]: Yeah, similar to what I alluded to above. We may want to break out these into into their own file in the `types` directory.

**Review by Subzeero — COMMENTED** [review]
Good to see all of this coming together!! I didn't get to take a look at this yesterday, but I thought it was worth it now, regardless.

**Pull request merged**
